prov/sm2/src/sm2.h

```diff
@@ -99,6 +99,12 @@ enum {
 
 #define SM2_ATOMIC_RMA_REQ (1 << 1)
 
+/* The SM2_UNEXP flag is used for unexpected receives. On the receiver, it
+indicates that the xfer_entry that the receiver is processing is actually in the
+xfer_ctx_pool and not in the receiver's freestack.
+*/
```

Comment:
Not sure what you mean here. The sender is as agnostic as possible to whether the receive is unexpected. If the sender has requested delivery complete, then the sender can't generate completion until the application has posted the receive and the receiver has copied the data into the buffer. So the sender needs to be aware of the unexpected receive in case of delivery complete semantics.

The unexpected flow is different for different protocols, so I didn't add all of it in this commit. I added comments for the specific protocols in those commits. See e.g. https://github.com/ofiwg/libfabric/blob/8e8a706a4e7d978b4a3e4a3b704b2659a5475f63/prov/sm2/src/sm2.h#L107-L112

Reply:
Delivery complete just means the data is somewhere in the receiver (whether the target buffer or a bounce buffer). For the case of an unexpected message, this means we can copy the payload into an internal buffer so when the receiver actually posts the buffer, the data is actually coming from an internal location rather than the command. This is how shm does it right now for inline, inject, and sar. The sender is completely unaware whether the data is in the posted receive or not.

Reply:
> Delivery complete just means the data is somewhere in the receiver (whether the target buffer or a bounce buffer).

That was not my understanding of delivery complete. While working on these changes, both @a-szegel  and I thought delivery complete meant that the data has to be in the receiver's application buffer. So this set of commits uses an unnecessarily higher standard and has quite a bit of logic to make that happen..

Are you OK with me just adding a comment about it for now?

Reply:
Yeah that's fine to just have a comment for now. Thanks!

Reply:
@sunkuamzn You might have mixed delivery complete with match complete.

Reply:
Yes I think we both made that mistake https://ofiwg.github.io/libfabric/main/man/fi_cq.3.html

Reply:
But the receiver's libfabric had to touch the message... it can't be sitting the the receivers fifo queue.  The receiver needs to do some processing, and return the FQE.